# Association Rules

#### Import libraries

In [1]:
import os
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder

### Example 1: Lecture

**Data from binary matrix in Excel**

<h4 style="color:blue"> Write Your Code Below: </h4>

In [ ]:
ex1_df = pd.read_excel(os.path.join('..', 'data', 'Association_Rules_Examples.xlsx'), 
                       sheet_name=1)

<h3 style="color:teal"> Expected Output: </h3>

Pizza  Burger   Soup  Salad  Fries  Garlic Bread   Soda  Water  \
tran_id                                                                   
1         True   False  False   True  False         False   True  False   
2        False    True  False  False  False         False   True  False   
3         True   False  False  False  False          True   True  False   
4        False    True  False  False   True         False  False   True   
5        False    True  False  False   True         False  False  False   
6         True   False  False  False  False         False   True  False   
7        False    True  False  False   True         False   True  False   
8        False   False   True   True  False         False  False   True   
9         True   False  False  False   True         False   True  False   
10        True   False  False   True  False         False   True  False   

         Ice Cream  
tran_id             
1            False  
2            False  
3            False  
4            False  
5             True  
6            False  
7            False  
8            False  
9            False  
10           False

**Display all frequent itemsets for a minimum support (`min_sup`) of 0.5.**

<h4 style="color:blue"> Write Your Code Below: </h4>

<h3 style="color:teal"> Expected Output: </h3>

support       itemsets
1      0.7         (Soda)
0      0.5        (Pizza)
2      0.5  (Soda, Pizza)

**Display strong association rules with a minimum confidence (`min_conf`) of 0.5 and only include the `antecedents`, `consequents`, `support`, `confidence`, and `lift` columns.**

<h4 style="color:blue"> Write Your Code Below: </h4>

In [ ]:
rule_cols = ['antecedents', 'consequents', 'support', 'confidence', 'lift']

<h3 style="color:teal"> Expected Output: </h3>

antecedents consequents  support  confidence      lift
1     (Pizza)      (Soda)      0.5    1.000000  1.428571
0      (Soda)     (Pizza)      0.5    0.714286  1.428571

### Example 2: Lecture

Array data structure

In [5]:
dataset = [['Laptop', 'Mouse', 'Keyboard'],
           ['Laptop', 'Mouse', 'Monitor'],
           ['Desk', 'Chair', 'Lamp'],
           ['Laptop', 'Desk', 'Mouse'],
           ['Mouse', 'Keyboard', 'Desk']]

Due to the size of this example data, a SparseDtype structure is not required. Check out the documentation for [pandas.sparseDtype](https://pandas.pydata.org/docs/reference/api/pandas.SparseDtype.html) for more information.

**Create a binary matrix for the dataset.**

<h4 style="color:blue"> Write Your Code Below: </h4>

<h3 style="color:teal"> Expected Output: </h3>

Chair   Desk  Keyboard   Lamp  Laptop  Monitor  Mouse
0  False  False      True  False    True    False   True
1  False  False     False  False    True     True   True
2   True   True     False   True   False    False  False
3  False   True     False  False    True    False   True
4  False   True      True  False   False    False   True

**Display all frequent itemsets for a minimum support (`min_sup`) of 0.6.**

<h4 style="color:blue"> Write Your Code Below: </h4>

<h3 style="color:teal"> Expected Output: </h3>

support         itemsets
2      0.8          (Mouse)
0      0.6           (Desk)
1      0.6         (Laptop)
3      0.6  (Mouse, Laptop)

**Display strong association rules with a minimum confidence (`min_conf`) of 0.5 and only include the `antecedents`, `consequents`, `support`, `confidence`, and `lift` columns.**

<h4 style="color:blue"> Write Your Code Below: </h4>

<h3 style="color:teal"> Expected Output: </h3>

antecedents consequents  support  confidence  lift
1    (Laptop)     (Mouse)      0.6        1.00  1.25
0     (Mouse)    (Laptop)      0.6        0.75  1.25

### Problem 14.2: Identifying Course Combinations

The Institute for Statistics Education at Statistics.com offers online courses in statistics and analytics, and is seeking information that will help in packaging and sequencing courses.  Consider the data in the file _CourseTopics.csv_, the first few rows of which are shown in the Table. These data are for purchases of online statistics courses at Statistics.com. Each row represents the courses attended by a single customer.
The firm wishes to assess alternative sequencings and bundling of courses. Use association rules to analyze these data, and interpret several of the resulting rules.

**Read in data from `coursetopics.csv` file.**

<h4 style="color:blue"> Write Your Code Below: </h4>

<h3 style="color:teal"> Expected Output: </h3>

Intro  DataMining  Survey  Cat Data  Regression  Forecast    DOE     SW
0   True        True   False     False       False     False  False  False
1  False       False    True     False       False     False  False  False
2  False        True   False      True        True     False  False   True
3   True       False   False     False       False     False  False  False
4   True        True   False     False       False     False  False  False

**Display all frequent itemsets for a minimum support (`min_sup`) of 0.01.**

<h4 style="color:blue"> Write Your Code Below: </h4>

<h3 style="color:teal"> Expected Output: </h3>

support                            itemsets
0   0.394521                             (Intro)
7   0.221918                                (SW)
3   0.208219                          (Cat Data)
4   0.208219                        (Regression)
2   0.186301                            (Survey)
..       ...                                 ...
58  0.010959      (DataMining, Survey, Forecast)
85  0.010959  (Intro, Regression, DOE, Cat Data)
86  0.010959   (Intro, Regression, SW, Cat Data)
57  0.010959    (Regression, DataMining, Survey)
70  0.010959           (Regression, Survey, DOE)

[89 rows x 2 columns]

**Display strong association rules with a minimum confidence (`min_conf`) of 0.1 and only include the `antecedents`, `consequents`, `support`, `confidence`, and `lift` columns.**

<h4 style="color:blue"> Write Your Code Below: </h4>

<h3 style="color:teal"> Expected Output: </h3>

antecedents             consequents   support  confidence  \
319        (Regression, SW)            (Intro, DOE)  0.019178    0.350000   
318            (Intro, DOE)        (Regression, SW)  0.019178    0.411765   
320       (Regression, DOE)             (Intro, SW)  0.019178    0.636364   
317             (Intro, SW)       (Regression, DOE)  0.019178    0.200000   
250  (Regression, Forecast)     (Intro, DataMining)  0.013699    0.357143   
247     (Intro, DataMining)  (Regression, Forecast)  0.013699    0.250000   

         lift  
319  7.514706  
318  7.514706  
320  6.636364  
317  6.636364  
250  6.517857  
247  6.517857

**Filter rules to have only one consequent.**

<h4 style="color:blue"> Write Your Code Below: </h4>

<h3 style="color:teal"> Expected Output: </h3>

antecedents   consequents   support  confidence  \
243    (Intro, Regression, Forecast)  (DataMining)  0.013699    0.714286   
262             (Intro, Survey, DOE)    (Cat Data)  0.010959    0.800000   
234    (Intro, DataMining, Cat Data)  (Regression)  0.016438    0.750000   
253        (Intro, Survey, Cat Data)    (Forecast)  0.013699    0.500000   
242  (Intro, Regression, DataMining)    (Forecast)  0.013699    0.500000   
313         (Intro, Regression, DOE)          (SW)  0.019178    0.777778   

         lift  
243  4.010989  
262  3.842105  
234  3.601974  
253  3.578431  
242  3.578431  
313  3.504801

**Interpret some of the rules.**

<h4 style="color:purple"> Write Your Free-Form Response Below: </h4>

- The first rule is "If Intro, Regression and Forecasting are taken, Data Mining is also taken." It has confidence of 71.4%, and a lift of 4.01.
- The second rule is "If Intro, Survey and DOE are taken, Categorical Data is also taken." It has confidence of 80% and lift of 3.84.

The support (a U c) for all rules is very low. under 4%. This means that the applicability of these rules is not great, and also that the chances are greater that we are not picking up true associations that will persist in the future -- just random noise.